### ISSUE TITLE
<touching (my self)> should give "false" instead of "true"

### ISSUE THREAD

**Issue by Costor** [jmoenig/Snap--Build-Your-Own-Blocks#1633]
I would like to propose that
![image](https://cloud.githubusercontent.com/assets/24916587/22175022/4e21ab5c-dfed-11e6-90f2-e39e1fdd1dbc.png)
should give "false" instead, and, as a result of this, I propose that
<touching (list of sprites)>  
should give "true" if and only if a different sprite touches "this" sprite. 

Currently <touching (list of sprites)> gives "true" already if "this" is in the (list of sprites).

All in all I believe that for most use cases, excluding "this" from touching itself would give more concise code and would avoid unexpected behaviour.

My reasoning is
a) there is no meaningful sense in which a collider can collide with itself, and in other game engines this is not considered. (Similarly, the < touching color > sensor does not automatically give "true" for colors that are in the costume of "this" sprite).
b) It is not intuitive: If you check <touching (list of sprites)> you almost always want to know whether "this" touches a different sprite. Game program logic almost always works with the idea of collider and collidee being different. 
c) It makes coding with lists of sprites complicated: checking "this" touching a (list of sprites), currently you always have to (1) make sure "this" sprite is not contained in the list, and if it is, (2) exclude it from the list before the touching check can meaningful be applied.  That takes an extra list copying operation and is expensive. (You could use the handy block (keep items such that [] from (list)) but that is the same complexity).
d) The other way round, it would be cheaper: To recreate today's behaviour it would just be  
< < (list of sprites) contains "this") >   OR  < touching (list of sprites) > > 

Kind regards, Costor

**Comment by cycomachead** [comment]
> a) there is no meaningful sense in which a collider can collide with itself, and in other game engines this is not considered. (Similarly, the < touching color > sensor does not automatically give "true" for colors that are in the costume of "this" sprite).

Not that I don't believe you -- but it would be helpful to see some documentation of this, particularly if it mentions why. 

Arguments that involve "there is no meaningful sense" "there is no reason" make me very skeptical. People come up with things they want to do that I think there is no reason for all the time. Sometimes there is a better way to accomplish something, but that doesn't mean there isn't always a reason. 


Given that collision is determined by what occupies that space on the stage, I find the current behavior understandable. I'm not a game programmer, so perhaps I'm in the minority here. 

I do completely see the use in b) (and c, d, but they all seem related), answering "am I touching anything other than myself", but to me a simple solution is to provide a list of "other sprites" or "sprites other than myself" or something. It seems to me that such a list is useful in a lot of senses because if you don't care about "touching myself" then a list of other sprites is useful when doing something like combining over AND and OR -- where in at one of those cases you'd need to filter out yourself from the list. (i.e. with the current behavior, it's the OR condition that needs filtering, but if we made this false we would make it harder to ask the question "am I touching all other sprites", which is something you could want to ask.)

So, I'm more curious than anything to hear what others think. I'm not opposed to this idea, but I do think there is a "less invasive" solution to problems you're experiencing.

**Comment by Costor** [comment]
I agree using "absolute" statements should be avoided (my fault - actually I copied this one in because of the notion of an collision which by the mere meaning of the word requires two parties - collider and collidee). Yes, there is no right or wrong here - its just about the design of < touching sprite >,  discussion points might be e.g. use cases, ease to express something in Snap, consistency in the approach etc.

**Comment by brianharvey** [comment]
+1 to disallow a sprite touching itself.

(a) In ordinary language, "touching myself" means you're doing something special, like feeling your pulse, or something else we don't have to go into here.  You're not "touching yourself" simply by virtue of occupying the same space as yourself.

(b) While in a sense it technically doesn't matter which way we decide because the user can always program the other option, it _does_ matter because ((myself) in front of (...)) is both easier to program and faster to execute than (keep items such that (not ((myself) = ( )))).

(c) And I agree that in actual programs, what the user wants to know is "am I touching anybody?" (to which the answer is currently always yes) or "whom am I touching?" (for which see point (a)).

**Comment by cycomachead** [comment]
> (a) In ordinary language, "touching myself" means you're doing something special, like feeling your pulse, or something else we don't have to go into here. You're not "touching yourself" simply by virtue of occupying the same space as yourself.

I thought about the human analogy but I don't really like it because touching is defined fairly specifically in that case--based on what your hands are doing - and if you asked if your hands are touching themselves - well I don't know how most people would answer, but to me, "yes" would be acceptable. 

> it does matter because ((myself) in front of (...)) is both easier to program and faster to execute than (keep items such that (not ((myself) = ( )))).

This is true -- but this is more about the API of the of `my` than the output of `touching ?`, no?

**Comment by brianharvey** [comment]
If you answered "yes" to "are your hands touching themselves" when they aren't, e.g., making a fist, everyone would assume you're a 10-year-old future mathematician making a joke, along the lines of "rather, rinse, repeat" being an infinite loop.

And, about the last point, as I said, there are lots of workarounds, and using (my (other sprites)) is one of them; but the _easiest_ way to think about the technical issue happens to agree with the _natural_ way to think about what the question means.

If I had my druthers, the TOUCHING? block would have "anybody" as one of the choices!  And it wouldn't always report true.  (And there'd be a TOUCHING WHOM reporter, too, so that it'd be fast to check.)  WHEN (TOUCHING? (ANYBODY)) ... TELL (TOUCHING WHOM) ...

**Comment by cycomachead** [comment]
I agree that touching (anybody)? shouldn't always report true (that would be useless) and that we should have that if possible -- but in my mind that is implicitly "anybody other than myself". 

I guess what I'm trying to get at is that if we do presume it's at least very unusual to ask the question "touching myself" then the problems arise from dealing with a list of sprites. It seems more clear, and precise, to me to have a list of sprites excluding myself. I guess what I'm arguing for is that regardless of whether or not we change touching myself that such a list is useful. 

-- 
Michael Ball
From my iPhone
michaelball.co

**Comment by brianharvey** [comment]
We already have (my (other sprites))!  That's not at issue.

**Comment by cycomachead** [comment]
Oh, well then in the case of needing to keep / add to lists, aren't all those problems moot?

And since we have this list, doesn't `touching my (other sprites)?` work as intended?

**Comment by brianharvey** [comment]
We have all been saying all along that it's possible to achieve the desired result either way.  The question is what the user expects to happen and how not to surprise the user unnecessarily.  If I'm remembering correctly from another thread, in the case of a sprite with (Scratch-style) clones, TOUCHING? of that sprite means touching any member of its clone family, but with a special case kludge that if the asker is one of the family it doesn't count.  The proposal here is that the degenerate case, if you will, of a clone family with zero clones should behave consistently with the case of a clone family with nonzero clones.

**Comment by cycomachead** [comment]
I get the proposal -- I'm saying that discussion around what the workarounds
are has been misleading to me. The work around seems to be to change the
option in the my block - not needing to bother with keep or in front of or
anything.  
  
  
I'm only really pushing back because a change such as this definitely has the
chance to break existing programs. I'd also like to understand, I guess from
Jens, if there was a particular reason Snap! works this way or if it's just by
nature of the way touching is implemented.

  

Michael Ball |
[michaelball.co](https://link.nylas.com/link/ec44e8vwbl1vjfjbdck69u19h/local-0
b4e0867-401f/0?redirect=http%3A%2F%2Fmichaelball.co&r=cmVwbHkrMDAxNmZhNzNkMjc4
NDhmMDdmODU5MmZiYTUwZGMwYTA1MjViNjhmMWVhNzZkOTYzOTJjZjAwMDAwMDAxMTQ5ZTQ4NzY5Mm
ExNjljZTBjMGYzMGNjQHJlcGx5LmdpdGh1Yi5jb20=)

![](https://link.nylas.com/open/ec44e8vwbl1vjfjbdck69u19h/local-0b4e0867-401f?
r=cmVwbHkrMDAxNmZhNzNkMjc4NDhmMDdmODU5MmZiYTUwZGMwYTA1MjViNjhmMWVhNzZkOTYzOTJj
ZjAwMDAwMDAxMTQ5ZTQ4NzY5MmExNjljZTBjMGYzMGNjQHJlcGx5LmdpdGh1Yi5jb20=)

  
On Jan 23 2017, at 2:40 pm, Brian Harvey <notifications@github.com> wrote:  

> We have all been saying all along that it's possible to achieve the desired
result either way. The question is what the user expects to happen and how not
to surprise the user unnecessarily. If I'm remembering correctly from another
thread, in the case of a sprite with (Scratch-style) clones, TOUCHING? of that
sprite means touching any member of its clone family, but with a special case
kludge that if the asker is one of the family it doesn't count. The proposal
here is that the degenerate case, if you will, of a clone family with zero
clones should behave consistently with the case of a clone family with nonzero
clones.

>

> —  
You are receiving this because you commented.  
Reply to this email directly, [view it on
GitHub](https://github.com/jmoenig/Snap--Build-Your-Own-Blocks/issues/1633#iss
uecomment-274641242&r=cmVwbHkrMDAxNmZhNzNkMjc4NDhmMDdmODU5MmZiYTUwZGMwYTA1MjVi
NjhmMWVhNzZkOTYzOTJjZjAwMDAwMDAxMTQ5ZTQ4NzY5MmExNjljZTBjMGYzMGNjQHJlcGx5LmdpdG
h1Yi5jb20=), or [mute the thread](https://github.com/notifications
/unsubscribe-auth/ABb6c2YhOkExNhfYwCf3_cc3J76o7aF-ks5rVSx2gaJpZM4LqGUL&r=cmVwb
HkrMDAxNmZhNzNkMjc4NDhmMDdmODU5MmZiYTUwZGMwYTA1MjViNjhmMWVhNzZkOTYzOTJjZjAwMDA
wMDAxMTQ5ZTQ4NzY5MmExNjljZTBjMGYzMGNjQHJlcGx5LmdpdGh1Yi5jb20=).![](https://git
hub.com/notifications/beacon/ABb6c2NzXsDjWN1b0g5hELmbMjdO0PL9ks5rVSx2gaJpZM4Lq
GUL.gif)

**Comment by brianharvey** [comment]
You're right, I'd forgotten about my(other sprites) when I said that about keep.

I'd really be interested to see a program broken by this change that isn't already standing on its head to work around the present behavior.

#threadHijack: We should really have some uniform system to make non-backwards-compatible changes in a way that remembers the change dates and remembers the save dates of projects and does some regular-expression-ish translation of the obsoleted block on load.

**Comment by Costor** [comment]
Would it be possible to just add an option to Snap (such as "thread safe scripts" or "turbo mode" on/off) so the user can choose? Eventually that option could also be switched on/off in a script (as "turbo mode"). 

(Btw - I feel the discussion would be simpler if the block would be renamed to  "am I overlapping with"  or "am I intersecting with" but that is impossible as it is a Scratch heritage and compatibility issue.)
